server/svix-server/src/core/cache/memory.rs

```diff
@@ -56,17 +56,52 @@ pub struct MemoryCache {
     map: SharedState,
 }
 
+impl MemoryCache {
+    async fn get_raw(&self, key: &str) -> Option<String> {
+        self.map
+            .read()
+            .await
+            .get(key)
+            .filter(|wrapper| check_is_expired(wrapper))
+            .map(|wrapper| wrapper.value.clone())
+    }
+
+    async fn set_raw(&self, key: &str, value: &str, ttl: Duration) -> Result<()> {
+        self.map
+            .write()
+            .await
+            .insert(String::from(key), ValueWrapper::new(value.to_string(), ttl));
+        Ok(())
+    }
+
+    async fn set_raw_if_not_exists(&self, key: &str, value: &str, ttl: Duration) -> Result<bool> {
+        let mut lock = self.map.write().await;
+
+        // TODO: use HashMap::try_insert when stable
+        // https://github.com/rust-lang/rust/issues/82766
+        if !lock.contains_key(key) {
+            lock.insert(key.to_owned(), ValueWrapper::new(value.to_owned(), ttl));
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+}
+
 #[async_trait]
 impl CacheBehavior for MemoryCache {
     async fn get<T: CacheValue>(&self, key: &T::Key) -> Result<Option<T>> {
-        Ok(self
-            .map
-            .read()
+        self.get_raw(key.as_ref())
             .await
-            .get(key.as_ref())
-            .filter(|wrapper| check_is_expired(wrapper))
-            .map(|wrapper| serde_json::from_str(&wrapper.value))
-            .transpose()?)
+            .map(|x| serde_json::from_str(&x).map_err(|e| e.into()))
+            .transpose()
+    }
+
+    async fn get_string<T: StringCacheValue>(&self, key: &T::Key) -> Result<Option<T>> {
+        self.get_raw(key.as_ref())
+            .await
+            .map(|x| x.try_into().map_err(|_| Error::DeserializationOther))
+            .transpose()
```

Comment:
If you add `get_raw`, `set_raw` and the likes to the trait, you can then save yourself from having to implement `get_string` and the likes in each backend, and just have a default implementation in the trait itself.
Essentially reducing all of this duplicate but identical code.

Does that make sense?